js/private/js_info.bzl

```diff
@@ -38,13 +38,31 @@ def js_info(
     Returns:
         A JsInfo provider
     """
+    # These are for legacy-users of js_info, who might pass in lists instead of depsets.
```

Comment:
Yeah so about this being a breaking change to a public API...

Thanks for including this for compatibility. I think it helps, but it's not sufficient. Old producing rules can work with newer rules_js, but old consuming rules cannot. I'm thinking through implications.

An org might have some custom rules in a private repo. A given version of those rules consumes only lists, another version only depsets. Then the end-user has to select the version of those rules that works with the rules_js version they install. Seems tricky to explain, and not obvious to users since the error message you'd get is just `Error: 'list' value has no field or method 'to_list'`.

The improvement to resource usage in this PR might justify the breaking nature of it. It's hard to say since we don't know how many usages of the public API there are.

Any ideas how to make it optional/less breaking?

Reply:
No worries, I wasn't sure the best way to introduce it across the board. Namely: if this would be something that we'd need to make in lock-step to all the repos at once, or if we make this change more incrementally. I can re-approach it with that in mind though!

Reply:
We don't really have the lock-step option anymore now that we shipped 1.0.0, at least in theory.

Reply:
So going over the codebases, I think all consumption of these providers from other Aspect repositories is currently through the `js_lib_helpers` struct. It looks like the main point of incompatibility is that those methods will return `depset`s instead of `list`s. So I've updated that API to use methods which can accept either lists or depsets, and return lists.

I don't know how y'all want to deal with the incompatibility, but my approach is aiming at:

1. Release a new `js_lib_depset_helpers` struct, and keep the `js_lib_helpers` struct for lists-users.
   - Repositories wanting the latest after this is added can change their code to work with `js_lib_depset_helpers` instead.
2. When y'all are ready to break things, make `js_lib_helpers` and `js_lib_depset_helpers` be the same struct.
3. And when you're ready to cleanup, remove `js_lib_depset_helpers`.

I did a GitHub-wide search for `*.bzl` using `JsInfo`, but only came across your repos, users of `bazelbuild/rules_nodejs`, and a `rivethealth/rules_javascript` (a separate attempt at JS rules for bazel), so this approach seems like it shouldn't cause too much tumult?

Reply:
That seems like a good compromise to me. Anyone who wrote their own rules in the last couple weeks and is using the helpers will be okay. Anyone interacting directly with the provider will have to figure out the version skew issue.

Of course the sooner we make and release the change the easier the migration is, since fewer things in the world will have taken a dependency on the lists version.